File: 2008/Qualification/A/code.cpp
#include <string>
#include <set>
#include <iostream>

int main(){ 
    std::string temp;
    std::getline(std::cin, temp);
    const int N = std::stoi(temp);

    for (int i = 1; i <= N; i++){ //Iterate through test cases
        std::cout << "Case #" << i << ": ";
        
        //Read in number of search engines and skip s lines so we can read in the
        //number of queries to process. We do not care what the different possible
        //engines are.  
        std::getline(std::cin, temp);
        auto s = std::stoi(temp); 
        auto j = s;
        while(j-- >= 0){
            std::getline(std::cin, temp);
        }
        auto q = std::stoi(temp);

        //Add each line to a set of queries to run. If adding the new query makes 
        //there be s distinct queries to run, then we had to run the unrun queries
        //before the current query is queued. Therefore we increment our changes,
        //and reset the set to only include the most recent query (the one we could
        //not queue with the rest).  
        int changes = 0;
        std::set<std::string> queries;
        while (q--){
            std::getline(std::cin, temp);
            queries.insert(temp);
            if (queries.size() == s){
                changes += 1;
                queries.clear();
                queries.insert(temp);
            }
        }

        std::cout << changes << "\n";
    } 
}

File: 2008/Qualification/A/py/code.py
import fileinput

def advance_iterable_n(iterable, n):
    while n > 0:
        n -= 1
        next(inFile)

def solve(inFile):
    #discard list of possible values
    numEngines = int(next(inFile))
    advance_iterable_n(inFile, numEngines)

    #iterate over queries, adding each one to a set until the set contains all
    #numEngines possible queries. Then increment count and reset set to only
    #include the newest query since we had to have cleared the queries before
    #it arrived.
    count = 0
    numQueries = int(next(inFile))
    currentBatch = set()
    while numQueries > 0 :
        numQueries -= 1
        temp = next(inFile)
        currentBatch.add(temp)
        if len(currentBatch) == numEngines:
            currentBatch = set([temp])
            count += 1
    return count

if __name__ == "__main__":
    inFile = fileinput.input()
    T = int(next(inFile))
    for i in range(1,T+1):
        print("Case #%d: %d" % (i, solve(inFile)))